substratevm/src/com.oracle.svm.agent.core/src/com/oracle/svm/agent/core/JNIHandleSetBase.java

```diff
@@ -0,0 +1,176 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.agent.core;
+
+import static com.oracle.svm.core.util.VMError.guarantee;
+import static com.oracle.svm.jni.JNIObjectHandles.nullHandle;
+
+import java.util.Arrays;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.graalvm.nativeimage.c.type.CTypeConversion;
+
+import com.oracle.svm.jni.JNIObjectHandles;
+import com.oracle.svm.jni.nativeapi.JNIEnvironment;
+import com.oracle.svm.jni.nativeapi.JNIMethodId;
+import com.oracle.svm.jni.nativeapi.JNIObjectHandle;
+
+/**
+ * Helps with creation and manages JNI handles for JVMTI agents.
+ *
+ * A JVMTI agent must provide a subclass of this class. It should contain handles to classes and
+ * methods that are needed across different JVMTI calls (for example, a JNIObjectHandle to
+ * java/lang/Class).
+ *
+ * For JNI handles that are created in a JVMTI callback and that should survive a return from native
+ * code, a helper method {@link #newTrackedGlobalRef} is provided.
+ *
+ * See also: {@link com.oracle.svm.agent.core.JvmtiAgentBase}.
+ */
+public abstract class JNIHandleSetBase {
+    private static final int INITIAL_GLOBAL_HANDLE_CAPACITY = 16;
+    private final ReentrantLock globalRefsLock = new ReentrantLock();
+    private JNIObjectHandle[] globalRefs = new JNIObjectHandle[INITIAL_GLOBAL_HANDLE_CAPACITY];
+    private int globalRefCount = 0;
+
+    JNIMethodId javaLangClassGetName;
+
+    public JNIHandleSetBase(JNIEnvironment env) {
+        JNIObjectHandle javaLangClass = findClass(env, "java/lang/Class");
+        try (CTypeConversion.CCharPointerHolder name = Support.toCString("getName"); CTypeConversion.CCharPointerHolder signature = Support.toCString("()Ljava/lang/String;")) {
+            javaLangClassGetName = Support.jniFunctions().getGetMethodID().invoke(env, javaLangClass, name.get(), signature.get());
+            guarantee(javaLangClassGetName.isNonNull());
+        }
+    }
+
+    /**
+     * Returns a local handle to a Java class object.
+     *
+     * If the class is not found, a RuntimeException is thrown. If the handle must survive a return
+     * from native code, {@link #newTrackedGlobalRef} should be used on the returned handle.
+     *
+     * @param env JNI environment of the thread running the JVMTI callback.
+     * @param className The VM type signature of the class.
+     * @return Local JNI handle representing the class object.
+     */
+    protected static JNIObjectHandle findClass(JNIEnvironment env, String className) {
+        try (CTypeConversion.CCharPointerHolder name = Support.toCString(className)) {
+            JNIObjectHandle h = Support.jniFunctions().getFindClass().invoke(env, name.get());
+            guarantee(h.notEqual(nullHandle()));
+            return h;
+        }
+    }
+
+    /**
+     * A convenience method to return a global handle to a Java class object.
```

Comment:
This documentation doesn't mention that the reference is freed eventually by this class and a user might believe they are responsible for freeing it. I would just reference method `newTrackedGlobalRef` here.

Reply:
I agree, will add a note for this.